### PR TITLE
[#213] [2/2 - Sample] Apply to use Flow for Compose and XML samples

### PR DESCRIPTION
**Pull request by luongvo**
Resolves https://github.com/nimblehq/android-templates/issues/213

## What happened 👀

- Apply to use [Flow](https://developer.android.com/kotlin/flow) to transfer the data from `Repository` through `UseCase` to `ViewModel`.
- Get rid of duplicating a new **try/catch** block in both `UseCase` and `Repository` when implementing https://github.com/nimblehq/android-templates/issues/212 as we can't access `retrofit objects` such as [HttpException](https://square.github.io/retrofit/2.x/retrofit/retrofit2/HttpException.html) in UseCase layer.
- Get rid of using a custom `UseCaseResult` to transfer data and exception from Usecase to ViewModel.
- Update all ViewModel logics and samples to adapt new structure.

## Insight 📝

- I don't think we should apply to use a custom `UseCaseResult` or even built-in [Result](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/), we can rely on kotlin.Flow to catch the error or collect data flow properly. 🤓 

## Proof Of Work 📹

The sample apps should work properly after migrating.

### CODE REVIEW

**Review by ryan-conway — APPROVED** [review]
👍🏻

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt
Comment (lydiasama) [inline template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: Shall we start the comment with Capital?
```suggestion
            // TODO: Do error mapping here
```

File: sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt
Comment (lydiasama) [inline sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt:19]: Shall we start the TODO comment with Capital?
```suggestion
            // TODO: Do error mapping here
```

File: sample-xml/data/src/main/java/co/nimblehq/sample/xml/data/repository/RepositoryImpl.kt
Comment (lydiasama) [inline sample-xml/data/src/main/java/co/nimblehq/sample/xml/data/repository/RepositoryImpl.kt:19]: Ditto here

**Review by lydiasama — COMMENTED** [review]
A few small comments.

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt
Comment (luongvo) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @lydiasama I think it's too strictly to follow that. I can't find any convention of Kotlin or Java requires to have `:` and start with a capital char. Let's make the dev's life easier 🙏 

https://www.jetbrains.com/help/idea/using-todo.html
https://developer.android.com/kotlin/style-guide#documentation
https://nimblehq.co/compass/development/code-conventions/kotlin/#comments

Comment (lydiasama) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: Then could you please update[ this TODO comment](https://github.com/nimblehq/android-templates/pull/292/files#diff-eb44200d9d6b03ba7a180a699c76534b141ad10297fb217d311a8fa4733a5a9dR13) to align with others?

Comment (luongvo) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @lydiasama It's not a big deal I think.
- It's not in my PR scope.
- We will remove it very soon.

Comment (lydiasama) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: > We will remove it very soon.

@luongvo  When will we remove it? Did we have a plan to continue it already?
And also, I feel it's not neat to have all comments that don't align with the same pattern. 🤔 So instead of updating the new comments that you added, you can update the old one to be consistent.

**Comment by chornerman** [comment]
@luongvo Double "to transfer" here 😄 

Screen Shot 2565-09-06 at 17 48 48

File: sample-xml/data/src/main/java/co/nimblehq/sample/xml/data/repository/RepositoryImpl.kt
Comment (chornerman) [inline sample-xml/data/src/main/java/co/nimblehq/sample/xml/data/repository/RepositoryImpl.kt:8]: Same here

**Review by chornerman — COMMENTED** [review]
The rest LGTM 🚀

File: sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt
Comment (luongvo) [reply sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt:8]: @chornerman they're different 🤔 1 is class, 1 is extension 😆

Comment (chornerman) [reply sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt:8]: Oh... my bad 😅

**Review by chornerman — APPROVED** [review]
🚀

**Review by AVI5HEK — APPROVED** [review]
LGTM 🚀

**Comment by ryan-conway** [comment]
@lydiasama Any more comments here? Would like to get your approval too before merging 😄

**Comment by lydiasama** [comment]
@ryan-conway https://github.com/nimblehq/android-templates/pull/292#discussion_r963500553

File: template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt
Comment (luongvo) [reply template/data/src/main/java/co/nimblehq/template/data/repository/RepositoryImpl.kt:19]: @lydiasama FYI I removed it here already https://github.com/nimblehq/android-templates/pull/311/files#diff-2ab311aa3921aeffb9dd9eb349a8ec774b2e63220b2793c49db434d68d0c52b2L19

File: sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt
Comment (luongvo) [reply sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt:19]: @lydiasama I will apply the same structure of error mapping in sample projects. Hence it will be removed soon in the [#212] Part 2/2.

**Comment by luongvo** [comment]
> @ryan-conway [#292 (comment)](https://github.com/nimblehq/android-templates/pull/292#discussion_r963500553)

@lydiasama Can I get your review again please?

File: sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt
Comment (lydiasama) [reply sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/RepositoryImpl.kt:19]: @luongvo Thanks for your answer. Then this is acceptable.

**Review by lydiasama — APPROVED** [review]

**Pull request merged**
